pkg/api/handlers/utils/containers.go

```diff
@@ -202,12 +203,29 @@ func waitRemoved(ctrWait containerWaitFn) (int32, error) {
 	return code, err
 }
 
-func waitNextExit(ctrWait containerWaitFn) (int32, error) {
-	_, err := ctrWait(define.ContainerStateRunning)
-	if err != nil {
-		return -1, err
-	}
-	return ctrWait(notRunningStates...)
+func waitNextExit(ctx context.Context, containerName string) (int32, error) {
+	runtime := ctx.Value("runtime").(*libpod.Runtime)
+	containerEngine := &abi.ContainerEngine{Libpod: runtime}
+	eventChannel := make(chan *events.Event)
+	errChannel := make(chan error)
+	opts := entities.EventsOptions{
+		EventChan: eventChannel,
+		Filter:    []string{"event=died", fmt.Sprintf("container=%s", containerName)},
+		Stream:    true,
+	}
+
+	// ctx is used to cancel event watching goroutine
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		errChannel <- containerEngine.Events(ctx, opts)
+	}()
+
+	evt, ok := <-eventChannel
+	if ok {
+		return int32(evt.ContainerExitCode), nil
+	}
+	return -1, <-errChannel
```

Comment:
Please, let's discuss the right approach before changing the code.

I still think we should return a custom error here. `errChannel` could return `nil` in line 228 and I believe that it must be an error that clearly identifies the (code) location.  The channel should not be closed since we're streaming, so we must receive the exit code of the container at some point.  To me, line 228 should not happen unless there's an internal error.

Reply:
> The channel should not be closed since we're streaming, so we must receive the exit code of the container at some point.

But it can be closed. It happens if `ctx` was canceled.

> To me, line 228 should not happen unless there's an internal error.

It can can happen and it's completely all right. It happens if `ctx` was canceled.

The `ctx` is canceled if request is canceled.
Example:
Your run `curl "http://localhost:1234/containers/${CTR}/wait?condition=next-exit" -v -XPOST`
but before the `curl` exits you pres `ctrl+c`.

Reply:
Thank you, yes that makes sense!  Could you add a code comment there explaining the reasoning?

Reply:
@vrothberg I added a comment before the final return, is it sufficient?